sphinx/ext/autodoc.py

```diff
@@ -508,6 +513,36 @@ def parse_name(self):
                         (self.objpath and '.' + '.'.join(self.objpath) or '')
         return True
 
+    def monkey_patch_import_function(self):
+        """Replace the builtin ``import`` function.
+
+        When a module cannot be imported, and its name begins by something it
+        ``autodoc_mock_imports``, return a dummy one which makes the
+        auto-documented modules "importable".
+        """
+        def _import(name, globals=None, locals=None, fromlist=(), level=None):
+            try:
+                if PY2:
+                    if globals is None:
+                        globals = {}
+                    if locals is None:
+                        locals = {}
+                    if level is None:
+                        level = -1
+                elif level is None:
+                    level = 0
+                return REALIMPORT(name, globals, locals, fromlist, level)
+            except ImportError:
+                if any(name.startswith(m)
+                       for m in self.env.config.autodoc_mock_imports):
+                    self.env.app.debug('[autodoc] adding a mock module %s!'
+                                       % name)
+                    return _MockModule(name)
+                else:
+                    raise
+
+        builtins.__import__ = _import
```

Comment:
I am not sure this is a well know way to do it, but it is the best I found to cover all problems.

Also, I like this approach since it is lazy. A module will be mocked only when required.